quickjs.c

```diff
@@ -54495,22 +54598,104 @@ uint32_t add_ic_slot(JSContext *ctx, JSInlineCache *ic, JSAtom atom, JSObject *o
     assert(cr != NULL);
     i = cr->index;
     for (;;) {
-        if (object->shape == cr->shape[i]) {
+        if (object->shape == cr->shape[i] && prototype == cr->proto[i]) {
             cr->prop_offset[i] = prop_offset;
             goto end;
         }
         i = (i + 1) % countof(cr->shape);
-        if (unlikely(i == cr->index))
+        if (unlikely(i == cr->index)) {
+            cr->index = (cr->index + 1) % countof(cr->shape);
             break;
+        }
     }
     sh = cr->shape[i];
+    if (cr->watchpoint_ref[i])
+        js_shape_delete_watchpoints(ctx->rt, sh, cr);
+    cr->prop_offset[i] = prop_offset;
     cr->shape[i] = js_dup_shape(object->shape);
     js_free_shape_null(ctx->rt, sh);
-    cr->prop_offset[i] = prop_offset;
+    if (prototype) {
+        JS_DupValue(ctx, JS_MKPTR(JS_TAG_OBJECT, prototype));
+        cr->proto[i] = prototype;
+        cr->watchpoint_ref[i] = js_shape_create_watchpoint(ctx->rt, cr->shape[i], cr, i,
+                          JS_DupAtom(ctx, atom));
+    }
+
 end:
     return ch->index;
 }
 
+int ic_watchpoint_delete_handler(JSRuntime* rt, JSInlineCacheRingSlot *ref, uint8_t offset, JSAtom atom, JSInlineCacheRingSlot *target) {
+    if (ref != target)
+        return 1;
+    assert(ref->proto[offset] != NULL);
+    JS_FreeValueRT(rt, JS_MKPTR(JS_TAG_OBJECT, ref->proto[offset]));
+    JS_FreeAtomRT(rt, atom);
+    ref->watchpoint_ref[offset] = NULL;
+    ref->proto[offset] = NULL;
+    ref->prop_offset[offset] = 0;
+    ref->shape[offset] = NULL;
+    return 0;
+}
+
+int ic_watchpoint_free_handler(JSRuntime* rt, JSInlineCacheRingSlot *ref, uint8_t offset, JSAtom atom) {
+    assert(ref->watchpoint_ref[offset] != NULL);
+    assert(ref->proto[offset] != NULL);
+    JS_FreeValueRT(rt, JS_MKPTR(JS_TAG_OBJECT, ref->proto[offset]));
+    JS_FreeAtomRT(rt, atom);
+    ref->watchpoint_ref[offset] = NULL;
+    ref->proto[offset] = NULL;
+    ref->prop_offset[offset] = 0;
+    ref->shape[offset] = NULL;
+    return 0;
+}
+
+int ic_delete_shape_proto_watchpoints(JSRuntime *rt, JSShape *shape, JSAtom atom) {
+    struct list_head *el, *el1;
+    JSObject *p;
+    JSInlineCacheRingSlot *cr;
+    JSShape *sh;
+    p = shape->proto;
+    while(likely(p)) {
+        if (p->shape->watchpoint)
+            list_for_each_safe(el, el1, p->shape->watchpoint) {
+                JSInlineCacheWatchpoint *o = list_entry(el, JSInlineCacheWatchpoint, link);
+                if (o->atom == atom) {
+                    cr = (JSInlineCacheRingSlot *)o->ref;
+                    sh = cr->shape[o->offset];
+                    ic_watchpoint_free_handler(rt, cr, o->offset, o->atom);
+                    js_free_shape_null(rt, shape);
+                    list_del(el);
+                    js_free_rt(rt, o);
+                }
+            }
```

Comment:
Braces for legibility and superfluous cast, ditto in ic_free_shape_proto_watchpoints. Change the return value to void if it's not returning anything interesting / you're not checking the return value at call sites.
```suggestion
        if (p->shape->watchpoint) {
            list_for_each_safe(el, el1, p->shape->watchpoint) {
                JSInlineCacheWatchpoint *o = list_entry(el, JSInlineCacheWatchpoint, link);
                if (o->atom == atom) {
                    cr = o->ref;
                    sh = cr->shape[o->offset];
                    ic_watchpoint_free_handler(rt, cr, o->offset, o->atom);
                    js_free_shape_null(rt, shape);
                    list_del(el);
                    js_free_rt(rt, o);
                }
            }
        }
```